set.c

```diff
@@ -0,0 +1,2149 @@
+}
+
+struct set_intersection_data {
+    VALUE set;
+    set_table *into;
+    set_table *other;
+};
+
+static int
+set_intersection_i(st_data_t key, st_data_t tmp)
+{
+    struct set_intersection_data *data = (struct set_intersection_data *)tmp;
+    if (set_lookup(data->other, key)) {
+        key = set_stringify(data->into, (VALUE)key);
+        RB_OBJ_WRITTEN(data->set, Qundef, (VALUE)key);
+        set_insert(data->into, key);
+    }
+
+    return ST_CONTINUE;
+}
+
+static VALUE
+set_intersection_block(RB_BLOCK_CALL_FUNC_ARGLIST(i, data))
+{
+    set_intersection_i((st_data_t)i, (st_data_t)data);
+    return i;
+}
+
+/*
+ *  call-seq:
+ *    set & enum -> new_set
+ *
+ *  Returns a new set containing elements common to the set and the given
+ *  enumerable object.
+ *
+ *    Set[1, 3, 5] & Set[3, 2, 1]             #=> #<Set: {3, 1}>
+ *    Set['a', 'b', 'z'] & ['a', 'b', 'c']    #=> #<Set: {"a", "b"}>
+ */
+static VALUE
+set_i_intersection(VALUE set, VALUE other)
+{
+    VALUE new_set = rb_funcall(rb_obj_class(set), id_new, 0);
+    set_table *stable = RSET_TABLE(set);
+    set_table *ntable = RSET_TABLE(new_set);
+
+    if (rb_obj_is_kind_of(other, rb_cSet)) {
+        set_table *otable = RSET_TABLE(other);
+        if (set_table_size(stable) >= set_table_size(otable)) {
+            /* Swap so we iterate over the smaller set */
+            otable = stable;
+            set = other;
+        }
+
+        struct set_intersection_data data = {
+            .set = new_set,
+            .into = ntable,
+            .other = otable
+        };
+        set_iter(set, set_intersection_i, (st_data_t)&data);
+    }
+    else {
+        struct set_intersection_data data = {
+            .set = new_set,
+            .into = ntable,
+            .other = stable
+        };
+        rb_block_call(other, enum_method_id(other), 0, 0, set_intersection_block, (VALUE)&data);
+    }
+
+    return new_set;
+}
+
+/*
+ *  call-seq:
+ *    include?(item) -> true or false
+ *
+ *  Returns true if the set contains the given object:
+ *
+ *    Set[1, 2, 3].include? 2   #=> true
+ *    Set[1, 2, 3].include? 4   #=> false
+ *
+ *  Note that <code>include?</code> and <code>member?</code> do not test member
+ *  equality using <code>==</code> as do other Enumerables.
+ *
+ *  This is aliased to #===, so it is usable in +case+ expressions:
+ *
+ *    case :apple
+ *    when Set[:potato, :carrot]
+ *      "vegetable"
+ *    when Set[:apple, :banana]
+ *      "fruit"
+ *    end
+ *    # => "fruit"
+ *
+ *  See also Enumerable#include?
+ */
+static VALUE
+set_i_include(VALUE set, VALUE item)
+{
+    return RBOOL(RSET_IS_MEMBER(set, item));
+}
+
+struct set_merge_args {
+  VALUE set;
+  set_table *into;
+};
+
+static int
+set_merge_i(st_data_t key, st_data_t data)
+{
+    struct set_merge_args *args = (struct set_merge_args *)data;
+    key = set_stringify(args->into, (VALUE)key);
+    RB_OBJ_WRITTEN(args->set, Qundef, (VALUE)key);
+    set_insert(args->into, key);
```

Comment:
```suggestion
    set_insert(args->into, key);
    RB_OBJ_WRITTEN(args->set, Qundef, (VALUE)key);
```

Reply:
Might actually be worth having a small helper to insert and then trigger the write barrier if the value was actually inserted, given how frequent that pattern is.